node_modules/oae-authz/lib/api.js

```diff
@@ -545,14 +564,141 @@ var _explodeGroupMemberships = function(principalId, callback) {
+    });
+
+};
+
+/**
+ * Get the full memberships AuthzGraph for a principal. The graph has the following guarantees:
+ *
+ *  * The nodes in the graph represent principals
+ *  * The resulting graph will always contain a node representing the given `principalId`, even if
+ *    they have no memberships
+ *  * Each node has an `id` property which indicates the principal id (e.g., "u:oae:branden",
+ *    "g:oae:developers")
+ *  * The inbound edges of each node denote a "has member" relationship
+ *      * i.e., (Branden) --> (OAE Developers) indicates that the "OAE Developers" group "has
+ *        member" "Branden"
+ *  * The outbound edges then obviously denote a "is member of" relationship
+ *      * i.e., (Branden) --> (OAE Developers) indicates that "Branden" "is member of" the "OAE
+ *        Developers" group
+ *  * Each edge has a `role` property which indicates the role the source principal has in the
+ *    target principal
+ *
+ * @param  {String}         principalId     The id of the principal whose full membership graph to get
+ * @param  {Function}       callback        Standard callback function
+ * @param  {Object}         callback.err    An error that occurred, if any
+ * @param  {AuthzGraph}     callback.graph  The memberships graph of the given principal
+ */
+var getPrincipalMembershipsGraph = module.exports.getPrincipalMembershipsGraph = function(principalId, callback) {
+    var validator = new Validator();
+    validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id provided'}).isPrincipalId();
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    }
+
+    // First try the full memberships cache
+    Cassandra.runAllPagesQuery('AuthzMembershipsCache', 'principalId', principalId, 'groupId', null, function(err, rows) {
+        if (err) {
+            return callback(err);
+        } else if (_.isEmpty(rows)) {
+            // If we have no cached memberships, go to the memberships CFs to try and resolve it
+            // recursively
+            return _explodeGroupMemberships(principalId, callback);
+        }
+
+        // We had some group memberships, so we construct the graph based on the encoded group
+        // members stored in the cache
+        var graph = new AuthzGraph();
+
+        // Seed the graph with the principal whose memberships graph we're building
+        graph.addNode(principalId);
+
+        // For every entry in the memberships cache, we need to extract its relevant members and
+        // create the membership relations
+        _.each(rows, function(row) {
+            // The id of a group that the `principalId` is directly or indirectly a member
+            var groupId = row.get('groupId').value;
+
+            // This is the stored members information for the group. It is of the form:
+            // `{'memberId': 'g:oae:another-group', 'role': 'manager'}`
+            var memberRoles = JSON.parse(row.get('value').value);
```

Comment:
We probably should try / catch this and complain very loudly if it ever errors out